packages/snaps-controllers/src/cronjob/CronjobController.ts

```diff
@@ -267,11 +340,126 @@ export class CronjobController extends BaseController<
   }
 
   /**
-   * Unregister all jobs related to the given snapId.
+   * Schedule a background event.
+   *
+   * @param backgroundEventWithoutId - Background event.
+   * @returns An id representing the background event.
+   */
+  scheduleBackgroundEvent(
+    backgroundEventWithoutId: Omit<BackgroundEvent, 'id' | 'scheduledAt'>,
+  ) {
+    const event = this.getBackgroundEventWithId(backgroundEventWithoutId);
+    event.scheduledAt = new Date().toISOString();
+    this.setUpBackgroundEvent(event);
+    this.update((state) => {
+      state.events[event.id] = castDraft(event);
+    });
+
+    return event.id;
+  }
+
+  /**
+   * Cancel a background event.
+   *
+   * @param id - The id of the background event to cancel.
+   * @throws If the event does not exist.
+   */
+  cancelBackgroundEvent(id: string) {
+    assert(
+      this.state.events[id],
+      `A background event with the id of "${id}" does not exist.`,
+    );
+
+    const timer = this.#timers.get(id);
+    timer?.cancel();
+    this.#timers.delete(id);
+    this.#snapIds.delete(id);
+    this.update((state) => {
+      delete state.events[id];
+    });
+  }
+
+  /**
+   * Assign an id to a background event.
+   *
+   * @param backgroundEventWithoutId - A background event with an unassigned id.
+   * @returns A background event with an id.
+   */
+  private getBackgroundEventWithId(
+    backgroundEventWithoutId: Omit<BackgroundEvent, 'id' | 'scheduledAt'>,
+  ): BackgroundEvent {
+    assert(
+      !hasProperty(backgroundEventWithoutId, 'id'),
+      `Background event already has an id: ${
+        (backgroundEventWithoutId as BackgroundEvent).id
+      }`,
+    );
+    const event = backgroundEventWithoutId as BackgroundEvent;
+    const id = this.generateBackgroundEventId();
+    event.id = id;
+    return event;
+  }
+
+  /**
+   * A helper function to handle setup of the background event.
+   *
+   * @param event - A background event.
+   */
+  private setUpBackgroundEvent(event: BackgroundEvent) {
+    const date = new Date(event.date);
+    const now = new Date();
+    const ms = date.getTime() - now.getTime();
+
+    const timer = new Timer(ms);
+    timer.start(() => {
+      this.executeBackgroundEvent(event).catch((error) => {
+        logError(error);
+      });
+
+      this.#timers.delete(event.id);
+      this.#snapIds.delete(event.id);
+      this.update((state) => {
+        delete state.events[event.id];
+      });
+    });
+
+    this.#timers.set(event.id, timer);
+    this.#snapIds.set(event.id, event.snapId);
+  }
+
+  /**
+   * Fire the background event.
+   *
+   * @param event - A background event.
+   */
+  private async executeBackgroundEvent(event: BackgroundEvent) {
+    await this.#messenger.call('SnapController:handleRequest', {
```

Comment:
Tech debt in this controller that we aren't just using `this.messagingSystem` 😄

Reply:
lol I guess I was trying to be overly neat